[0035] Rework elementwise operations (#651)
microsoft/hlsl-specs: proposals/0035-linalg-matrix.md

scope matrix, and must be called in uniform control flow when loading a wave
scope matrix.

#### Matrix::Length

```c++
template <MatrixScope ScopeLocal = Scope>
typename hlsl::enable_if<Scope != MatrixScope::Thread && ScopeLocal == Scope,
                         uint>::type
Matrix::Length();
```

Returns the number of matrix components accessible to the current thread. The
mapping and distribution of threads to matrix elements is opaque and

> implementation-specific. The value returned by `Length` may not be the same
> across all threads. The sum of the values returned by `Length` across all

Reviewer: nit: I think the positive is easier to parse

```suggestion
implementation-specific. The value returned by `Length` may be different
for each thread. The sum of the values returned by `Length` across all
```

scope matrix, and must be called in uniform control flow when loading a wave
scope matrix.

#### Matrix::Length

```c++
template <MatrixScope ScopeLocal = Scope>
typename hlsl::enable_if<Scope != MatrixScope::Thread && ScopeLocal == Scope,
                         uint>::type
Matrix::Length();
```

> Returns the number of components stored in the thread's thread local memory.

Reviewer 2: Consider the following:

```md
Returns the number of matrix components accessible to the current thread. The
mapping and distribution of threads to matrix elements is opaque and
implementation-specific. The value returned by `Length` may not be the same
across all threads. The sum of the values returned by `Length` across all
threads must be greater than or equal to the total number of matrix elements.
Some implementations may map multiple threads to the same matrix element.
Therefore, developers should take this into consideration when programming
side-effects, such as atomic operations and/or UAV writes, within user-defined
matrix operations.

May be called from non-uniform control flow. However, given the above rules,
calling `Length` from divergent threads may result in unpredictable behavior.
For example, the number of matrix elements accessible to each thread will be
inconsistent across different implementations.
```

#### Matrix::GetCoordinate

```c++
template <MatrixScope ScopeLocal = Scope>
typename hlsl::enable_if<Scope != MatrixScope::Thread && ScopeLocal == Scope,
                         uint2>::type Matrix::GetCoordinate(uint);
```

Converts an index into the thread-local collection of matrix elements to the
matrix column and row returned as a `uint2`. The mapping of thread local indices

> to matrix coordinates is implementation defined.

Reviewer 2: Consider the following:

```md
Converts a specified index into row and column coordinates.  The valid
range of `Index` is `[0, Length()-1]`.  If the value of `Index` is out of
range, then the result value is undefined.
```

#### Matrix::Get

```c++
template <MatrixScope ScopeLocal = Scope>
typename hlsl::enable_if<Scope != MatrixScope::Thread && ScopeLocal == Scope,
                           ElementType>::type Matrix::Get(uint);
```

Gets the element corresponding to the thread-local index provided. The mapping

> of thread local indices to matrix coordinates is implementation defined.

Reviewer 2: Consider the following:

```md
Retrieves the value of a matrix component at the specified index.  The valid
range of `Index` is `[0, Length()-1]`.  If the value of `Index` is out of
range, then the result value is undefined.
```

#### Matrix::Set

```c++
template <MatrixScope ScopeLocal = Scope>
typename hlsl::enable_if<Scope != MatrixScope::Thread && ScopeLocal == Scope,
                         void>::type Matrix::Set(uint, ElementType);
```

Sets the element at the corresponding thread-local index to the provided value.
The mapping of thread local indices to matrix coordinates is implementation

> defined.

Reviewer 2: Consider the following:

```md
Sets the value of a matrix component at the specified index.  The valid
range of `Index` is `[0, Length()-1]`.  If the value of `Index` is out of
range, then the result value is undefined.
```

  // Element-wise operations
  template <typename T>
  typename hlsl::enable_if<hlsl::is_arithmetic<T>::value &&
                               Scope != MatrixScope::Thread,
                           Matrix>::type
  operator+=(T V) {
    for (uint I = 0; I < Length(); ++I)
      Set(I, Get(I) + V);
    return this;
  }

  template <typename T>
  typename hlsl::enable_if<hlsl::is_arithmetic<T>::value &&
                               Scope != MatrixScope::Thread,
                           Matrix>::type
  operator-=(T V) {
    for (uint I = 0; I < Length(); ++I)
      Set(I, Get(I) - V);
    return this;
  }

  template <typename T>
  typename hlsl::enable_if<hlsl::is_arithmetic<T>::value &&
                               Scope != MatrixScope::Thread,
                           Matrix>::type
  operator*=(T V) {
    for (uint I = 0; I < Length(); ++I)
      Set(I, Get(I) * V);
    return this;
  }

  template <typename T>
  typename hlsl::enable_if<hlsl::is_arithmetic<T>::value &&
                               Scope != MatrixScope::Thread,
                           Matrix>::type
  operator/=(T V) {
    for (uint I = 0; I < Length(); ++I)
      Set(I, Get(I) / V);
    return this;

>   }

Reviewer 3: I expect you're aware, but for the sake of those who may not be, these operators are not supported by HLSL2021.

Author: Yes, and with the current API shape we really have two viable options. We could just not include these APIs (since they're trivial to implement), or we could propose relaxing the overload restrictions in HLSL 202x.
